site/public/js/chatroom.js

```diff
@@ -0,0 +1,284 @@
+    if (confirm(`This will delete chatroom '${chatroom_name}'. Are you sure?`)) {
+        const url = `${base_url}/delete`;
+        const fd = new FormData();
+        fd.append('csrf_token', csrfToken);
+        fd.append('chatroom_id', chatroom_id);
+        $.ajax({
+            url: url,
+            type: 'POST',
+            data: fd,
+            processData: false,
+            cache: false,
+            contentType: false,
+            success: function (data) {
+                try {
+                    const msg = JSON.parse(data);
+                    if (msg.status !== 'success') {
+                        console.error(msg);
+                        window.alert('Something went wrong. Please try again.');
+                    }
+                    else {
+                        window.location.reload();
+                    }
+                }
+                catch (err) {
+                    console.error(err);
+                    window.alert('Something went wrong. Please try again.');
+                }
+            },
+            error: function (err) {
+                console.error(err);
+                window.alert('Something went wrong. Please try again.');
+            },
+        });
+    }
+}
+
+function toggle_chatroom(chatroomId, active) {
+    const form = document.getElementById(`chatroom_toggle_form_${chatroomId}`);
+    if (active && confirm('This will terminate this chatroom session. Are you sure?')) {
+        form.submit();
+    }
+    else {
+        form.submit();
+    }
+}
+
+function showJoinMessage(message) {
+    const toast = document.querySelector('.chatroom-toast');
+    toast.textContent = message;
+    toast.style.visibility = 'visible';
+    toast.style.opacity = '0.85';
+    setTimeout(() => {
+        toast.style.opacity = '0'; // fade out
+    }, 3000);
+}
+
+function handleChatOpen(msg) {
+    const tableBody = document.querySelector('#chatrooms-table tbody');
+    if (!tableBody) {
+        return;
+    }
+    if (document.getElementById(`chatroom-row-${msg.id}`)) {
+        return;
+    }
+    const tr = document.createElement('tr');
+    tr.id = `chatroom-row-${msg.id}`;
+
+    tr.innerHTML = `
+        <td>
+            <span class="display-short" title="${msg.title}">
+                ${msg.title.length > 30 ? `${msg.title.slice(0, 30)}...` : msg.title}
+            </span>
+        </td>
+        <td>
+            ${msg.host_name}
+        </td>
+        <td>
+            <span class="display-short" title="${msg.description}">
+                ${msg.description.length > 45 ? `${msg.description.slice(0, 45)}...` : msg.description}
+            </span>
+        </td>
+        <td>
+            <a href="${msg.base_url}/${msg.id}" class="btn btn-primary">Join</a>
+            ${msg.allow_anon ? `<i> or </i><a href="${msg.base_url}/${msg.id}/anonymous" class="btn btn-default">Join As Anon.</a>` : ''}
+        </td>
+    `;
```

Comment:
I placed the do not merge label because of this comment. Please do not self-resolve this comment and verify with me that it is properly addressed.

There is an XSS security vuln here. Need to do this functionality without using `.innerHTML`. Some good options are to manually create the elements in JS and use that to inject in (can be a bit of a pain) or to use something like Vue.js. See calendar.js as an example for the first one.

Reply:
Yes this should be safe. Thanks for addressing.